clients/drcachesim/tests/trace_interval_analysis_unit_tests.cpp

```diff
@@ -149,63 +340,193 @@ class test_analysis_tool_t : public analysis_tool_t {
+        return true;
+    }
+    int
+    get_outstanding_snapshot_count()
+    {
+        return outstanding_snapshots_;
+    }
 
 private:
     int seen_memrefs_;
-    int seen_parallel_memrefs_;
+    std::vector<recorded_snapshot_t> expected_state_snapshots_;
+    int outstanding_snapshots_;
+
+    // Data tracked per shard.
+    struct per_shard_t {
+        memref_tid_t shard_id_;
+        uintptr_t magic_num_;
+        int seen_memrefs_;
+    };
 
-    static constexpr uintptr_t kShardData = 0x8badf00d;
+    static constexpr uintptr_t kMagicNum = 0x8badf00d;
+    static constexpr memref_tid_t kInvalidTid = -1;
 };
 
 static bool
 test_non_zero_interval(bool parallel)
 {
     constexpr uint64_t kIntervalMicroseconds = 100;
     std::vector<memref_t> refs = {
-        gen_instr(1, 1), gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 1), gen_instr(1, 2),
-        gen_data(1, true, 100, 4), gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 50),
-        gen_instr(1, 3),
-        // 0th interval ends here.
-        gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 101), gen_instr(1, 4),
-        // 1st interval ends here.
-        gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 490), gen_exit(1)
-        // 4th interval ends here.
+        // Trace for a single worker which has two constituent shards. (scheduler_t
+        // does not guarantee that workers will process shards one after the other.)
+        // Expected active interval_id: shard_1_local - shard_2_local - whole_trace
+        gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 40),  // 0 - _ - 0
+        gen_instr(1, 10000),                             // 0 - _ - 0
+        gen_data(1, true, 1234, 4),                      // 0 - _ - 0
+        gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 151), // _ - 0 - 1
+        gen_instr(2, 20000),                             // _ - 0 - 1
+        gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 170), // 1 - _ - 1
+        gen_instr(1, 10008),                             // 1 - _ - 1
+        gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 201), // 2 - _ - 2
+        gen_instr(1, 20004),                             // 2 - _ - 2
+        gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 210), // _ - 1 - 2
+        gen_instr(2, 20008),                             // _ - 1 - 2
+        gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 270), // _ - 1 - 2
+        gen_instr(2, 20008),                             // _ - 1 - 2
+        gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 490), // _ - 3 - 4
+        gen_instr(2, 20012),                             // _ - 3 - 4
+        gen_marker(1, TRACE_MARKER_TYPE_TIMESTAMP, 590), // 5 - _ - 5
+        gen_exit(1),                                     // 5 - _ - 5
+        gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 610), // _ - 5 - 6
+        gen_instr(2, 20016),                             // _ - 5 - 6
+        gen_exit(2)                                      // _ - 5 - 6
+    };
+    // Format for interval_switch_point: <tid, seen_memrefs, interval_id>
```

Comment:
What is interval_switch_point?  Not seeing it elsewhere

Reply:
I meant interval_end_point_t. Forgot to change here after I renamed the struct in my local changes. Fixed.